WordPress/Classes/Login/WordPressDotComAuthenticator.swift

```diff
@@ -0,0 +1,107 @@
+import AuthenticationServices
+import Foundation
+import UIKit
+
+import Alamofire
+
+/// Log in or sign up a WordPress.com account via web.
+///
+/// API doc: https://developer.wordpress.com/docs/oauth2/
+struct WordPressDotComAuthenticator {
+    enum Error: Swift.Error {
+        case invalidCallbackURL
+        case obtainAccessToken
+        case cancelled
+    }
+
+    func authenticate(from viewController: UIViewController) async throws -> String {
+        let clientId = ApiCredentials.client
+        let clientSecret = ApiCredentials.secret
+        let redirectURI = "x-wordpress-app://oauth2-callback"
+
+        let authorizeURL = URL(string: "https://public-api.wordpress.com/oauth2/authorize")!
+            .appending(queryItems: [
+                URLQueryItem(name: "client_id", value: clientId),
+                URLQueryItem(name: "redirect_uri", value: redirectURI),
+                URLQueryItem(name: "response_type", value: "code"),
+                URLQueryItem(name: "scope", value: "global"),
+            ])
+
+        let callbackURL: URL = try await withCheckedThrowingContinuation { continuation in
+            DispatchQueue.main.async {
+                let provider = ContextProvider(window: viewController.view.window ?? UIWindow())
+                let session = ASWebAuthenticationSession(url: authorizeURL, callbackURLScheme: "x-wordpress-app") { url, error in
+                    if let url {
+                        continuation.resume(returning: url)
+                    } else {
+                        DDLogWarn("Error from authentication session: \(String(describing: error))")
+                        continuation.resume(throwing: Error.cancelled)
+                    }
+                }
+                session.presentationContextProvider = provider
+                session.start()
+            }
+        }
+
+        return try await handleAuthorizeCallbackURL(callbackURL, clientId: clientId, clientSecret: clientSecret, redirectURI: redirectURI)
+    }
+
+    private func handleAuthorizeCallbackURL(
+        _ url: URL,
+        clientId: String,
+        clientSecret: String,
+        redirectURI: String
+    ) async throws -> String {
+        guard let query = URLComponents(url: url, resolvingAgainstBaseURL: true)?.queryItems,
+              let code = query.first(where: { $0.name == "code" })?.value else {
+            throw Error.invalidCallbackURL
+        }
+
+        var tokenRequest = URLRequest(url: URL(string: "https://public-api.wordpress.com/oauth2/token")!)
+        tokenRequest.httpMethod = "POST"
+        let parameters = [
+            "grant_type": "authorization_code",
+            "client_id": clientId,
+            "client_secret": clientSecret,
+            "redirect_uri": redirectURI,
+            "code": code,
+        ]
+
+        do {
+            tokenRequest = try URLEncodedFormParameterEncoder().encode(parameters, into: tokenRequest)
```

Comment:
I think it's okay to depend on that? As far as I know, we are not planning on removing Alamofire. Also, is there an alternative at the moment?

I didn't use `URL.query` because I believe there are some subtle differences between query and webform. i.e. `+` is allowed in query, but is decoded as space in webform.

From https://url.spec.whatwg.org/#urlencoded-parsing:
> Replace any 0x2B (+) in name and value with 0x20 (SP).

Reply:
I'm OK merging with it since there is seemingly no alternative. 
I'd personally like to remove `AlamofireImage` since the app has its own `ImageDownloader` and caches, but `Alamofire` itself offers quite a bit of value. I won't advocate for removing it.